File: 714.买卖股票的最佳时机含手续费.py
#
# @lc app=leetcode.cn id=714 lang=python3
#
# [714] 买卖股票的最佳时机含手续费
#

# @lc code=start
class Solution:
    def maxProfit(self, prices, fee):
        n = len(prices)
        min_price = prices[0]
        profit = 0
        for i in range(1, n):
            if prices[i] < min_price:
                min_price = prices[i]
            if prices[i] >= min_price and prices[i] <= min_price+fee:
                continue
            if prices[i] > min_price+fee:
                profit += prices[i] - min_price - fee
                min_price = prices[i] - fee  # 继续持有，相当于反向加上每一天的手续费
        return profit


if __name__ == '__main__':
    prices = [1,3,2,8,4,9]
    fee = 2
    print(Solution().maxProfit(prices, fee))
# @lc code=end

File: 714.买卖股票的最佳时机含手续费.java
/*
 * @lc app=leetcode.cn id=714 lang=java
 *
 * [714] 买卖股票的最佳时机含手续费
 */

// @lc code=start
class Solution {
    public int maxProfit(int[] prices, int fee) {
        // 动态规划
        if (prices.length == 0) {
            return 0;
        }

        int[][] profits = new int[prices.length][2];
        profits[0][0] = -prices[0]-fee;
        profits[0][1] = 0;

        for (int i = 1; i < prices.length; i++) {
            profits[i][0] = Math.max(profits[i-1][0], profits[i-1][1]-prices[i]-fee);
            profits[i][1] = Math.max(profits[i-1][1], profits[i-1][0]+prices[i]);
        }
        return profits[prices.length-1][1];

    }
}
// @lc code=end

